_posts/2021-02-12-the-best-test-is-the-test-you-dont-have-to-write.md

```diff
@@ -0,0 +1,192 @@
+We sometimes have to pause and reflect on what models we're using, what business
+feature we're implementing and use code to write this knowledge down.
+
+Now that we know our request can be in 4 possibles states, how does it look like
+in code?
+
+Such a mindset is something we're more used to see in codebases written with
+ML-languages like Haskell:
+
+```haskell
+data Request = NotAsked
+             | Loading
+             | Failure Error
+             | Success TheData
+```
+
+But nothing prevent us to write something similar with more classic languages,
+here Typescript: 
+
+```typescript
+type Request<E, P> = 
+  | { _type: 'NotAsked' }
+  | { _type: 'Loading' }
+  | { _type: 'Failure', error: E } 
+  | { _type: 'Success', payload: P }
+```
+
+We can clearly see the benefit here. The request can't be at the same time in
+loading and with an error. 
+
+Working with such model is both easy and secure: we're forced to handle each
+state to access the one with data but we're guided by the 4 states defined
+earlier:
+
+```tsx
+function Component() {
+  const request = useMyRequest()
+
+  switch (request._type) {
+    case 'NotAsked':
+      return <NotAskedScreen />
+
+    case 'Loading':
+      return <LoadingScreen />
+
+    case 'Failure':
+      return <ErrorScreen error={request.error} />
+
+    case 'Success':
+      return <SuccessScreen data={request.payload} />
+  }
+}
+```
+
+We have covered all valid states and no impossible ones. The developper writing
+the `useMyRequest` hook has 4 less tests to write. The one using the hook, being
+unsure of the quality of the work of his colleague, has also 4 tests less to
+write. Everyone is happy and can come home early. Good work!
+
+## The tools
+
+They are tools that can help us writing such kind of code.
+
+Typed language, first. They are not mandatory but they certainly can help a
+developper write and use such code more easily.
+
+Mathematics, with the type theory and especially [Alebraic Data
+Type](https://en.wikipedia.org/wiki/Algebraic_data_type) (ADT). The `Request`
+type we written just above is an ADT. 
+
+Libraries that leverage thoses concepts and functional programing.  Don't bother
+writing the `Request` by hand.  [`RemoteData` is a
+library](https://github.com/devexperts/remote-data-ts) that does just that.
+[`fp-ts` also offer a lot of ADTs](https://github.com/gcanti/fp-ts) to model
+what's going on in our codebase.
+
+Last but not least, our brains! It's not just about using technical and
+low-level concepts such as `Request`, `Option`, `Task`. Is also about using the
+same mindset to reprent *business* states the codebase should be in.
+
+## Conclusion 
+
+Does it mean we should stop writing tests? Certainly not. Tests are absolutly
+necessary to prevent our codebase to fail.
+
+But remember, we need tests because our code can fail. Reducing the number of
+way our code can fail should be our priority as much as writing tests for when
+it can.
```

Comment:
```suggestion
But remember, we need tests because our code can fail. Reducing the number of
way our code can fail should be our priority. We can then writing tests for the few remaining chances of failure.
```